### PR TITLE
FlutterProject refactoring and test coverage

### PR DESCRIPTION
**Pull request by mravn-google**
Avoid separate types for `android/` and `.android/` sub-projects. Similarly for iOS.

### CODE REVIEW

File: packages/flutter_tools/lib/src/project.dart [hunk 1]
```diff
@@ -167,33 +151,52 @@ class IosProject {
...
+    return isModule
+        ? directory.childDirectory('Flutter').childDirectory('FlutterPluginRegistrant')
+        : directory.childDirectory('Runner');
+  }
 }
 
-/// Represents the contents of the android/ folder of a Flutter project.
+/// Represents the Android sub-project of a Flutter project.
+///
+/// Instances will reflect the contents of the `android/` sub-folder of
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/project.dart]: What will happen after running `flutter materialize`? I thought that was the whole reason to have two types from the beginning.
I guess the different subdirectories can be made to point into `android/blah/` and `.android/foo` respectively, but `AndroidProject.directory` becomes ambiguous.
But then again, perhaps we never use it on the outside (in that case it should be made private).

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 2]
```diff
@@ -200,22 +200,26 @@ distributionUrl=https\\://services.gradle.org/distributions/gradle-$gradleVersio
...
-/// If [requireSdk] is `true` this will fail with a tool-exit if no Android Sdk
+/// If [requireSdk] is `true` this will fail with a tool exit if no Android Sdk
 /// is found.
 Future<void> updateLocalProperties({
   @required FlutterProject project,
   BuildInfo buildInfo,
   bool requireAndroidSdk = true,
 }) async {
+  if (project.manifest == null) {
+    throwToolExit('Invalid `pubspec.yaml`');
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Perhaps we should never be able to create a FlutterProject with a null manifest?

File: packages/flutter_tools/lib/src/project.dart [hunk 3]
```diff
@@ -73,84 +82,59 @@ class FlutterProject {
...
-  bool get hasExampleApp => _exampleDirectory.childFile('pubspec.yaml').existsSync();
+  /// True, if this project has an example application.
+  bool get hasExampleApp => _exampleDirectory(directory).existsSync();
 
   /// The directory that will contain the example if an example exists.
-  Directory get _exampleDirectory => directory.childDirectory('example');
+  static Directory _exampleDirectory(Directory directory) => directory.childDirectory('example');
+
+  /// The directory that contains the Android app to be launched by `flutter run`.
+  Directory get _androidDirectory => directory.childDirectory(isModule ? '.android' : 'android');
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/project.dart]: What about moving this logic to the AndroidProject class? - it is anyway passed isModule.

File: packages/flutter_tools/lib/src/project.dart [hunk 4]
```diff
@@ -73,84 +82,59 @@ class FlutterProject {
...
+
+  /// The directory that contains the iOS app to be launched by `flutter run`.
+  Directory get _iosDirectory => directory.childDirectory(isModule ? '.ios' : 'ios');
 
   /// Generates project files necessary to make Gradle builds work on Android
   /// and CocoaPods+Xcode work on iOS, for app and module projects only.
   Future<void> ensureReadyForPlatformSpecificTooling() async {
-    if (!directory.existsSync() || hasExampleApp) {
+    if (!directory.existsSync() || hasExampleApp)
       return;
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/project.dart]: Move return to previous line?

**Review by sigurdm — APPROVED** [review]
Generally LGTM if you have a plan in mind for the `flutter materialize` case.

**Comment by mravn-google** [comment]
Thanks, yes. My plan is to encapsulate the knowledge of which parts are in `android/app`, `.android/app`, and `.android/Flutter` inside the `AndroidProject` class. As you note, that will mean encapsulating also the `directory` getter.

File: packages/flutter_tools/lib/src/project.dart (hunk 4 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/project.dart]: I believe I'm following the style guide here: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#separate-the-if-expression-from-its-statement

File: packages/flutter_tools/lib/src/project.dart (hunk 3 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/project.dart]: `AndroidProject` and `IosProject` now take the parent `FlutterProject` as a constructor parameter instead.

File: packages/flutter_tools/lib/src/project.dart (hunk 1 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/project.dart]: Yes, I'll work towards encapsulating it.

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 2 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/android/gradle.dart]: Right. I'll push that to a separate PR.

**Pull request merged**